### PR TITLE
Add UNet

### PR DESCRIPTION
**Pull request by kevinyamauchi**
This PR adds a unet for semantic segmentation as a pytorch lightning model. We can add a dataset and some example code as a follow-up PR.

### CODE REVIEW

**Comment by kevinyamauchi** [comment]
@alisterburt and @lorenzlamm, I removed python 3.11 as it was causing issues with torch/monai. I think this is fine, as 3.11 is pretty bleeding edge at this point. What do you think?

Otherwise, I think this is ready for review!

File: src/membrain_seg/networks/unet.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    """
+
+    def __init__(
+        self,
+        spatial_dims: int = 3,
+        in_channels: int = 1,
+        out_channels: int = 1,
+        channels: Tuple[int, ...] = (16, 32, 64, 128, 256),
+        strides: Tuple[int, ...] = (2, 2, 2, 2),
+        num_res_units=2,
```
Comment (alisterburt) [inline src/membrain_seg/networks/unet.py:58]:
```suggestion
        num_res_units=2,
```

not in parameters in docstring - not 100% what this refers to without delving into the monai implementation

**Review by alisterburt — APPROVED** [review]
gorgeous - implementation looks great, very clear! Couple minor comments but otherwise good to go

qs around monai:
- is it stable/likely to change (should we consider vendoring?)
- is it a clean dep to bring in?

File: src/membrain_seg/networks/unet.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        """Implementation of the forward pass.
+
+        See the pytorch-lightning module documentation for details.
+        """
+        return self._model(x)
+
+    def configure_optimizers(self):
+        """Set up the Adam optimzier.
+
+        See the pytorch-lightning module documentation for details.
```
[the remark this answers is not in the document]
Comment (kevinyamauchi) [reply src/membrain_seg/networks/unet.py:111]: these spaces are required by the pre-commit

[the remark this answers is not in the document]
Comment (alisterburt) [reply src/membrain_seg/networks/unet.py:111]: Sorry not sure I understand - I didn't remove the line between, just added a link to the docs

File: src/membrain_seg/networks/unet.py (hunk 1 above)
Comment (kevinyamauchi) [reply src/membrain_seg/networks/unet.py]: I added a docstring

**Comment by kevinyamauchi** [comment]
Thanks for the review @alisterburt ! The line breaks you flagged are required by our pre-commit. I would suggest that we leave them as is and the re-vist the pre-commit separately if you prefer a different style.

> qs around monai:
> 
>     is it stable/likely to change (should we consider vendoring?)

In my experience, it is stable. I would recommend that we keep the dependency and vendor only if necessary.

>     is it a clean dep to bring in?

yeah, as specified, it will only bring in numpy and torch (which we will depend on anyway here)

https://github.com/Project-MONAI/MONAI/blob/9fd6d4ced3efb344cf3b24c51d899d8fafc882d5/setup.cfg#L44-L46

File: src/membrain_seg/networks/unet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevinyamauchi) [reply src/membrain_seg/networks/unet.py:113]: ah sorry -totally misread the comment. I'll add your suggestion.

**Comment by alisterburt** [comment]
excellent! Not sure what you mean by line breaks exactly (I only added a link after the existing doc) but it's a minor thing, let's merge!

**Pull request merged**